lua/nvim-treesitter/textobjects.lua

```diff
@@ -66,13 +67,158 @@ function M.select_textobject(query_string)
+                          bufnr,
+                          "yes, set cursor!")
+    end
+  end
+end
+
+function M.swap_next(query_string)
+  swap_textobject(query_string, 1)
+end
+
+function M.swap_previous(query_string)
+  swap_textobject(query_string, -1)
+end
+
+function M.goto_adjacent(query_string, forward, start, same_parent, overlapping_range_ok)
+  local bufnr, _, node = textobject_at_point(query_string)
+  local adjacent_textobject
+  if forward then
+    adjacent_textobject = M.next_textobject(node,  query_string, same_parent, overlapping_range_ok, bufnr)
+  else
+    adjacent_textobject = M.previous_textobject(node,  query_string, same_parent, overlapping_range_ok, bufnr)
+  end
+
+  if adjacent_textobject then
+    utils.set_jump()
+
+    local adjacent_textobject_range = {adjacent_textobject:range()}
+    if start then
+      api.nvim_win_set_cursor(api.nvim_get_current_win(),
+                              { adjacent_textobject_range[1] + 1, adjacent_textobject_range[2] })
+    else
+      api.nvim_win_set_cursor(api.nvim_get_current_win(),
+                              { adjacent_textobject_range[3] + 1, adjacent_textobject_range[4] })
+    end
+  end
+end
+
+-- luacheck: push ignore 631
+M.goto_next_start = function(query_string) M.goto_adjacent(query_string, 'forward', 'start', not 'same_parent', 'overlap ok') end
+M.goto_next_end = function(query_string) M.goto_adjacent(query_string, 'forward', not 'start', not 'same_parent', 'overlap ok') end
+M.goto_previous_start = function(query_string) M.goto_adjacent(query_string, not 'forward', 'start', not 'same_parent', 'overlap ok') end
+M.goto_previous_end = function(query_string) M.goto_adjacent(query_string, not 'forward', not 'start', not 'same_parent', 'overlap ok') end
+-- luacheck: pop
+
+function M.next_textobject(node, query_string, same_parent, overlapping_range_ok, bufnr)
+  local node = node or ts_utils.get_node_at_cursor()
+  local bufnr = bufnr or api.nvim_get_current_buf()
+  if not node then return end
+
+  local _, _, node_end = node:end_()
+  local search_start, _
+  if overlapping_range_ok then
+    _, _, search_start = node:start()
+  else
+    _, _, search_start = node:end_()
+  end
+  local scoring_function = function(match)
+                              if match.node == node then return end
+                              if not same_parent or node:parent() == match.node:parent() then
+                                local _, _, start = match.node:start()
+                                local _, _, end_ = match.node:end_()
+                                return start > search_start and end_ >= node_end
+                              end
+                            end
+  local filter_function = function(match)
+                            local _, _, node_start = match.node:start()
+                            return -node_start
+                          end
```

Comment:
i'm not too found of this kind of code because of this weird indentation but we'll keep it like this for the moment.

Reply:
Will fix it. Probably as a local function